src/database-controller/src/poller/index.js

```diff
@@ -0,0 +1,131 @@
+const config = require('@dbc/poller/config');
+const fetch = require('node-fetch');
+const { deleteFramework } = require('@dbc/core/k8s');
+// maxPending is set to 1 to avoid the queue to be too long.
+// If any framework is not synced/deleted, it will be synced/deleted in the next polling round.
+const lock = new AsyncLock({ maxPending: 1 });
+const databaseModel = new DatabaseModel(
+  config.dbConnectionStr,
+  config.maxDatabaseConnection,
+);
+
+async function mockDeleteEvent(snapshot) {
+  await fetch(`${config.writeMergerUrl}/api/v1/watchEvents/DELETED`, {
+    method: 'POST',
+    body: snapshot.getString(),
+    headers: { 'Content-Type': 'application/json' },
+    timeout: config.writeMergerConnectionTimeoutSecond * 1000,
+  });
+}
+
+function deleteHandler(snapshot, pollingTs) {
+  const frameworkName = snapshot.getName();
+  logger.info(
+    `Will delete framework ${frameworkName}. PollingTs=${pollingTs}.`,
+  );
+  lock
+    .acquire(frameworkName, async () => {
+      try {
+        // We only delete framework here, ignoring the job add-ons.
+        // Because most job add-ons are patched in the creation time, so they will by deleted automatically.
+        // If some add-ons are not successfully patched, they will be deleted by the watch dog service.
+        await deleteFramework(snapshot.getName());
+        logger.info(
+          `Framework ${frameworkName} is successfully deleted. PollingTs=${pollingTs}.`,
+        );
+      } catch (err) {
+        if (err.response && err.response.statusCode === 404) {
+          // for 404 error, mock a delete to write merger
+          logger.warn(
+            `Cannot find framework ${frameworkName} in API Server. Will mock a deletion to write merger.`,
+          );
+          await mockDeleteEvent(snapshot);
+        } else {
+          // for non-404 error
+          throw err;
+        }
+      }
+    })
+    .catch(err => {
+      logger.error(
+        `An error happened when delete framework ${frameworkName} and pollingTs=${pollingTs}:`,
+        err,
+      );
+    });
+}
+
+function synchronizeHandler(snapshot, addOns, pollingTs) {
+  const frameworkName = snapshot.getName();
+  logger.info(
+    `Start synchronizing request of framework ${frameworkName}. PollingTs=${pollingTs}`,
+  );
+  lock
+    .acquire(frameworkName, async () => {
+      await synchronizeRequest(snapshot, addOns);
+      logger.info(
+        `Request of framework ${frameworkName} is successfully synchronized. PollingTs=${pollingTs}.`,
+      );
+    })
+    .catch(err => {
+      logger.error(
+        `An error happened when synchronize request for framework ${frameworkName} and pollingTs=${pollingTs}:`,
+        err,
+      );
+    });
+}
+
+async function poll() {
+  const pollingTs = new Date().getTime();
+  try {
+    logger.info(`Start polling. PollingTs=${pollingTs}`);
+    const frameworks = await databaseModel.Framework.findAll({
+      attributes: [
+        'name',
```

Comment:
Have you tested the poller when our fast path fail?

Reply:
Tested.